xclim/indices/_agro.py

```diff
@@ -1067,17 +1069,185 @@ def _get_rain_season(pram):
+        has already been applied input dataset and will skip the resampling step.
+    window : int
+        Averaging window length relative to the resampling frequency. For example, if `freq="MS"`,
+        i.e. a monthly resampling, the window is an integer number of months.
+    dist : str
+        Name of the univariate distribution.
+        (see :py:mod:`scipy.stats`).
+    method : str
+        Name of the fitting method, such as `ML` (maximum likelihood), `APP` (approximate). The approximate method
+        uses a deterministic function that doesn't involve any optimization.
+    offset: Quantified
+        Distributions bounded by zero (e.g. "gamma", "fisk") can be used for datasets with negative values by using a offset:
+        `da + offset`
+    indexer
+        Indexing parameters to compute the indicator on a temporal subset of the data.
+        It accepts the same arguments as :py:func:`xclim.indices.generic.select_time`.
+    """
+    # "WPM" method doesn't seem to work for gamma or pearson3
+    dist_and_methods = {"gamma": ["ML", "APP"], "fisk": ["ML", "APP"]}
+    if dist not in dist_and_methods:
+        raise NotImplementedError(f"The distribution `{dist}` is not supported.")
+    if method not in dist_and_methods[dist]:
+        raise NotImplementedError(
+            f"The method `{method}` is not supported for distribution `{dist}`."
+        )
+
+    if offset:
+        with xarray.set_options(keep_attrs=True):
+            da = da + convert_units_to(offset, da, context="hydro")
+
+    da, group = _preprocess_standardized_index(da, freq, window, **indexer)
+
+    def wrap_fit(da):
+        if indexer != {} and da.isnull().all():
+            select_dims = {d: 0 if d != "time" else [0, 1] for d in da.dims}
+            fitted = fit(da[select_dims], dist, method)
+            return fitted.broadcast_like(da.isel(time=0, drop=True))
+        return fit(da, dist, method)
+
+    cal_range = (
+        da.time.min().dt.strftime("%Y-%m-%d"),
+        da.time.max().dt.strftime("%Y-%m-%d"),
+    )
+    params = da.groupby(group).map(wrap_fit)
+    params.attrs = {
+        "Calibration period": cal_range,
+        "freq": freq,
+        "window": window,
+        "scipy_dist": dist,
+        "method": method,
+        "group": group,
+        "time_indexer": indexer,
+        "units": "",
+        "offset": offset,
+    }
+    return params
+
+
+def _get_standardized_index(da, params, **indexer):
+    scipy_dist = get_dist(params.attrs["scipy_dist"])
+    group_name = params.attrs["group"].rsplit(".")[-1]
+    param_names = params.dparams.values
+
+    def wrap_cdf_ppf(da, params, group_idxs):
+        if indexer != {} and np.isnan(da).all():
+            return np.zeros_like(da) * np.NaN
+
+        spi = np.zeros_like(da)
+        # loop over groups (similar to np.groupies idea, but np.groupies can't be used with custom functions)
+        for ii, group_idx in enumerate(list(dict.fromkeys(group_idxs).keys())):
```

Comment:
I think this can be done, yes. It would just require a tad bit more attention, to make sure that arrays are well sorted. But I think this was already a pre-requesite, I will double check, then 

```python
 for group_idx in (list(dict.fromkeys(group_idxs).keys()):
            pars = params[group_idx - 1]
```
or something even simpler for the loop should do the trick

Reply:
I thought params` keys where the group index.